cmd/dep/status.go

```diff
@@ -335,6 +388,142 @@ func (cmd *statusCommand) validateFlags() error {
 	return nil
 }
 
+// OldStatus contains information about all the out of date packages in a project.
+type OldStatus struct {
+	ProjectRoot string
+	Constraint  gps.Constraint
+	Revision    gps.Revision
+	Latest      gps.Version
+}
+
+type rawOldStatus struct {
+	ProjectRoot, Constraint, Revision, Latest string
+}
+
+func (os OldStatus) getConsolidatedConstraint() string {
+	var constraint string
+	if os.Constraint != nil {
+		if v, ok := os.Constraint.(gps.Version); ok {
+			constraint = formatVersion(v)
+		} else {
+			constraint = os.Constraint.String()
+		}
+	}
+	return constraint
+}
+
+func (os OldStatus) getConsolidatedLatest(revSize uint8) string {
+	latest := ""
+	if os.Latest != nil {
+		switch revSize {
+		case shortRev:
+			latest = formatVersion(os.Latest)
+		case longRev:
+			latest = os.Latest.String()
+		}
+	}
+	return latest
+}
+
+func (os OldStatus) marshalJSON() *rawOldStatus {
+	return &rawOldStatus{
+		ProjectRoot: os.ProjectRoot,
+		Constraint:  os.getConsolidatedConstraint(),
+		Revision:    string(os.Revision),
+		Latest:      os.getConsolidatedLatest(longRev),
+	}
+}
+
+func (cmd *statusCommand) runOld(ctx *dep.Ctx, out oldOutputter, p *dep.Project, sm gps.SourceManager) error {
+	// While the network churns on ListVersions() requests, statically analyze
+	// code from the current project.
+	ptree, err := p.ParseRootPackageTree()
+	if err != nil {
+		return err
+	}
+
+	// Set up a solver in order to check the InputHash.
+	params := gps.SolveParameters{
+		ProjectAnalyzer: dep.Analyzer{},
+		RootDir:         p.AbsRoot,
+		RootPackageTree: ptree,
+		Manifest:        p.Manifest,
+		// Locks aren't a part of the input hash check, so we can omit it.
+	}
+
+	// Check update for all the projects.
+	params.ChangeAll = true
+
+	solver, err := gps.Prepare(params, sm)
+	if err != nil {
+		return errors.Wrap(err, "fastpath solver prepare")
+	}
+
+	solution, err := solver.Solve(context.TODO())
+	if err != nil {
+		return errors.Wrap(err, "runOld")
+	}
+
+	var oldStatuses []OldStatus
+	lockProjects := p.Lock.Projects()
+	solutionProjects := solution.Projects()
+
+	sort.Slice(solutionProjects, func(i, j int) bool {
+		return solutionProjects[i].Ident().Less(solutionProjects[j].Ident())
+	})
+
+	sort.Slice(lockProjects, func(i, j int) bool {
+		return lockProjects[i].Ident().Less(lockProjects[j].Ident())
+	})
+
+	for i := range solutionProjects {
+		lProj := lockProjects[i]
+		sProj := solutionProjects[i]
```

Comment:
Thinking about it now, I feel doubtful about this optimization. What if an update includes a new dependency,(a transitive dependency), in that case, the solution lock would have the new project and would be different from the existing lock, resulting failing this optimization and wrong results. What do you think?